File: backend/src/main/webapp/frontend/ts/picture/figure/FigureAggregator.ts
import {Figure} from "./Figure.js";

export abstract class FigureAggregator extends Figure {
    protected constructor(
        type: string,
        public readonly children: Array<Figure>
    ) {
        super(type)
    }

    draw(ctx: CanvasRenderingContext2D) {
        this.children.forEach(child => child.draw(ctx))
    }
}

File: backend/src/main/java/ru/vityaman/itmo/web/lab/servlets/backend/model/figure/FigureAggregator.java
package ru.vityaman.itmo.web.lab.servlets.backend.model.figure;

import ru.vityaman.itmo.web.lab.servlets.backend.model.Hashable;

import java.util.List;
import java.util.stream.Collectors;

public abstract class FigureAggregator extends Figure {
    private final List<? extends Figure> children;

    FigureAggregator(String type, List<? extends Figure> children) {
        super(type);
        this.children = children;
    }

    public final List<? extends Figure> children() {
        return children;
    }

    @Override
    public final String hash() {
        return Hashable.hash(
            children().stream()
                .map(Hashable::hash)
                .collect(Collectors.joining(":", type(), ""))
        );
    }
}
